### DISCUSSION TITLE
Adding a response header in a dependency after yield

### DISCUSSION THREAD

**Discussion by imbolc** [fastapi/fastapi#7965] [Questions] [answered]
I'm looking for a way to automatically save cookie based session. I can use a starlette middleware for this but it implies some overhead for endpoints which don't use session. So I tried to use a yielded dependency:
```python
async def get_session(request: Request, response: Response):
    session = await load_session(request)
    yield session
    response.set_cookie("session", session.encode())
```
But `set_cookie` doesn't work this way as request is already sent when it calls.

Do you have any suggestions?

**Comment by trim21** [comment]
dependency are called before handler called, so there's no way to get response in args. 

Maybe we could consider pass it by `response = yield obj` in fastapi (won't work now)?

**Comment by dmontagu** [comment]
Actually, there are some mechanisms for updating the response from dependencies (e.g., you can add response headers in a dependency). If you can determine the cookie at the start of the request it may be possible to get this to work with minor changes to fastapi (basically just support for setting cookies on the response from a dependency; maybe this is even already supported).

But as of now, contextmanager dependencies are fundamentally incapable of modifying the response *after* execution of the endpoint. There's not a clear way, for example, to apply dependencies to the response in a consistent order in general using FastAPI's dependency injection framework (if there were multiple response-modifying dependencies). This is really the job of middleware.

You can essentially get custom middleware on a route-by-route basis by subclassing `APIRoute` following [this docs page](https://fastapi.tiangolo.com/tutorial/custom-request-and-route/), but note this is a relatively advanced topic. It may be better to just use a middleware for this if possible; the added overhead should be minimal.

**Comment by imbolc** [comment]
Updating cookies from dependencies before (or without) yield works. But not after yield, eg in the following example only `before=yield` cookie will be set:
```python
def set_cookie(response: Response):
    response.set_cookie("before", "yield")
    yield
    response.set_cookie("after", "yield")

@app.get("/")
def index(_=Depends(set_cookie)):
    return {}
```

> There's not a clear way, for example, to apply dependencies to the response in a consistent order in general using FastAPI's dependency injection framework

Yeah, I was thinking about this too, so dependencies can't reliable replace middlewares, but still it will work for some cases. I guess I close this issue and open another one as a feature request. Thanks, guys :)

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue @imbolc  :+1:

**Comment by tiangolo** [comment, marked as answer]
An additional response here with an example.

You cannot set data after the `yield` in a dependency with `yield`, not right now (I haven't found a use case for it), it doesn't get access to the  response object after `yield`, it's intended to be used only to close resources, at least that's what has been.

If you want to refresh a cookie, you could do that in dependencies alone, and set the cookie in the same dependency using the response parameter. The cookies from that response parameter are copied to the final response returned to the client (unless you return a response directly).

So, you could for example have this auto-refresh cookie in a dependency, this includes some basic way to store data in a cookie, a dependency to verify the data, one endpoint that reads (and triggers the cookie auto refresh), and a login endpoint just to set the initial cookie:

```Python
import base64
import json
from typing import Annotated, Any

from fastapi import Cookie, Depends, FastAPI, Form, HTTPException, Response
from pydantic import BaseModel, ValidationError

app = FastAPI()


class CookieData(BaseModel):
    username: str
    count: int


def auto_refresh_cookie(
    response: Response, session: Annotated[str | None, Cookie()] = None
) -> CookieData:
    if session is None:
        raise HTTPException(status_code=401, detail="Session cookie not found")

    try:
        data = CookieData.model_validate_json(base64.b64decode(session).decode("utf-8"))
    except (json.JSONDecodeError, ValueError, ValidationError):
        raise HTTPException(status_code=401, detail="Invalid session cookie")
    data.count += 1
    new_session = base64.b64encode(data.model_dump_json().encode("utf-8")).decode(
        "utf-8"
    )
    response.set_cookie(key="session", value=new_session, httponly=True, secure=True)
    return data


@app.get("/")
def root(cookie: Annotated[CookieData, Depends(auto_refresh_cookie)]) -> dict[str, Any]:
    return {"message": f"Hello, {cookie.username}!", "count": cookie.count}


@app.post("/login")
def login(username: Annotated[str, Form()], response: Response) -> dict[str, Any]:
    data = CookieData(username=username, count=0)
    session = base64.b64encode(data.model_dump_json().encode("utf-8")).decode("utf-8")
    response.set_cookie(key="session", value=session, httponly=True, secure=True)
    return {"message": "Logged in successfully"}
```

If you login with user "charlie", and then you the main endpoint a few times, you will get responses:

```
{
  "message": "Hello, charlie!",
  "count": 1
}
```

then:

```
{
  "message": "Hello, charlie!",
  "count": 2
}
```

and so on.
